src/crimson/os/seastore/journal/circular_bounded_journal.cc

```diff
@@ -502,19 +502,24 @@ CircularBoundedJournal::write_ertr::future<>
 CircularBoundedJournal::write_header()
 {
   LOG_PREFIX(CircularBoundedJournal::write_header);
-  ceph::bufferlist bl;
-  try {
-    bl = encode_header();
-  } catch (ceph::buffer::error &e) {
-    DEBUG("unable to encode header block from underlying deivce");
-    return crimson::ct_error::input_output_error::make();
-  }
-  ceph_assert(bl.length() <= get_block_size());
-  DEBUG(
-    "sync header of CircularBoundedJournal, length {}",
-    bl.length());
-  assert(device);
-  return device_write_bl(device->get_journal_start(), bl);
+  return seastar::do_with(
+    bufferptr(ceph::buffer::create_page_aligned(get_block_size())),
+    [this, FNAME](auto &bp) {
+    ceph::bufferlist bl = encode_header();
+    ceph_assert(bl.length() <= get_block_size());
+    DEBUG(
+      "sync header of CircularBoundedJournal, length {}",
+      bl.length());
+    assert(device);
+    auto iter = bl.begin();
+    assert(bl.length() < get_block_size());
+    iter.copy(bl.length(), bp.c_str());
+    return device->write(device->get_journal_start(), bp
```

Comment:
`bufferptr` should already support reference counter, so no need to use do-with if the write parameter is not a reference `bufferptr&`, also refer to `Segment::write()`.

Reply:
I also encountered memory leak even after removing do-with without `bufferptr&`. This is the same as what write_superblock() in BlockSegmentManager did. Am I missing something?

Reply:
The caller doesn't need to worry about do-with each time, if the device implementation can handle the lifecycle correctly.

For example, segment_manager still needs to extend the buffer lifecycle internally using do-with for the asynchronous dma_write():
https://github.com/ceph/ceph/blob/01ad87ef30f99cae14baa152cc0b7bdf6ec0a114/src/crimson/os/seastore/segment_manager/block.cc#L100-L114